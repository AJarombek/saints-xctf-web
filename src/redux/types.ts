/**
 * Types for the Redux store.
 * @author Andrew Jarombek
 * @since 7/23/2020
 */

export type RootState = {
    auth: AuthState,
    logs: LogsState
}

export type LogsState = {
    isFetching: boolean;
    didInvalidate: boolean;
    lastUpdated: number;
    items: Logs;
    feeds: LogFeeds;
    newComments: NewComments;
}

export type AuthState = {
    auth: Auth;
    user: Users;
}

export interface Meta {
    isFetching?: boolean;
    lastUpdated?: number;
    didInvalidate?: boolean;
    serverError?: string;
}

export type Auth = {
    isFetching?: boolean;
    lastUpdated?: number;
    signedInUser?: string;
    status?: string;
}

export type Users = {
    [key: string]: UserMeta;
}

export interface UserMeta extends User, Meta {}

export interface User {
    activation_code?: string;
    class_year?: number;
    deleted?: string;
    description?: string;
    email?: string;
    favorite_event?: string;
    first?: string;
    last?: string;
    last_signin?: string;
    location?: string;
    member_since?: string;
    password?: string;
    salt?: string;
    subscribed?: string;
    username?: string;
    week_start?: string;
}

enum LogType { RUN = "run", BIKE = "bike", SWIM = "swim", OTHER = "other" }
enum Metric { MILES = "miles", KILOMETERS = "kilometers", METERS = "meters" }

export type Log = {
    log_id: number;
    username: string;
    first: string;
    last: string;
    name: string;
    date: string;
    type: LogType;
    distance?: number;
    metric?: Metric;
    miles?: number;
    time?: string;
    pace?: string;
    feel: number;
    location?: string;
    description?: string;
    timeCreated?: string;
    comments?: Comment[];
}

export type Logs = {
    [key: string]: Record<string, Log>;
}

export type LogFeeds = {
    [key: string]: {
        filterBy: string;
        bucket: string;
        pages: LogFeedPages
    };
}

export type LogFeedPages = {
    [key: string]: LogFeedPage
}

export type LogFeedPage = {
    isFetching: boolean;
    lastUpdated: number;
    items: Log[];
    serverError: string;
}

export type Comment = {
    comment_id: number;
    username: string;
    first: string;
    last: string;
    log_id: number;
    time: string;
    content: string;
}

export type NewComments = {
    [key: string]: {
        isFetching: boolean;
        lastUpdated: number;
        created?: boolean;
        serverError?: string;
    }
}
