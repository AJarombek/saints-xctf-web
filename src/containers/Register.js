/**
 * Register component for people to create new accounts.
 * @author Andrew Jarombek
 * @since 5/9/2020
 */

import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { useHistory } from 'react-router-dom';

import { userAuthenticated } from '../utils/auth';
import NavBar from '../components/shared/NavBar';
import RegisterBody from '../components/register/RegisterBody';
import { signIn } from '../redux/modules/auth';
import {
  registerPersonalInfo,
  registerCredentials,
  registerBack
} from '../redux/modules/registration';

const mapStateToProps = state => ({
  auth: state.auth.auth,
  user: state.auth.user,
  registration: state.registration
});

const mapDispatchToProps = {
  registerPersonalInfo,
  registerCredentials,
  registerBack
};

const Register = ({ auth = {}, user = {}, registration = {}, registerPersonalInfo,
                    registerCredentials, registerBack }) => {
  const { signedIn } = auth;
  const { stage = 0 } = registration;

  const history = useHistory();

  useEffect(() => {
    if (userAuthenticated(user, signedIn)) {
      history.push('/dashboard');
    }
  }, [user]);

  return (
    <div className="sxctf-register">
      <NavBar includeHeaders={["home", "signIn", "logo"]}/>
      <RegisterBody
        stage={stage}
        registerPersonalInfo={registerPersonalInfo}
        registerCredentials={registerCredentials}
        registerBack={registerBack}
        registration={registration}
      />
    </div>
  );
};

Register.propTypes = {
  auth: PropTypes.shape({
    isFetching: PropTypes.bool,
    lastUpdated: PropTypes.number,
    signedIn: PropTypes.bool,
    status: PropTypes.string
  }),
  user: PropTypes.shape({
    isFetching: PropTypes.bool,
    didInvalidate: PropTypes.bool,
    lastUpdated: PropTypes.number,
    activation_code: PropTypes.string,
    class_year: PropTypes.number,
    deleted: PropTypes.string,
    description: PropTypes.string,
    email: PropTypes.string,
    favorite_event: PropTypes.string,
    first: PropTypes.string,
    last: PropTypes.string,
    last_signin: PropTypes.string,
    location: PropTypes.string,
    member_since: PropTypes.string,
    password: PropTypes.string,
    subscribed: PropTypes.string,
    username: PropTypes.string,
    week_start: PropTypes.string
  }),
  registration: PropTypes.shape({
    isFetching: PropTypes.bool,
    lastUpdated: PropTypes.number,
    valid: PropTypes.bool,
    status: PropTypes.string,
    stage: PropTypes.number,
    first: PropTypes.string,
    last: PropTypes.string,
    email: PropTypes.string
  }),
  registerPersonalInfo: PropTypes.func.isRequired,
  registerCredentials: PropTypes.func.isRequired,
  registerBack: PropTypes.func.isRequired
};

export default connect(mapStateToProps, mapDispatchToProps)(Register);
