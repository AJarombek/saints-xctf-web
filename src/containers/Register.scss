/*
 * Sass styling for the Register component
 * @author Andrew Jarombek
 * @since 5/9/2020
 */

@import "../mixins";

.sxctf-register {
  background-color: #f5f5f5;
  @include lightNavDarkDropdown;
}
