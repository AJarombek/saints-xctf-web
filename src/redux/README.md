### Overview

Configure Redux as the state management system for the application.

### Files

| Filename                 | Description                                                                |
|--------------------------|----------------------------------------------------------------------------|
| `modules`                | Redux modules containing Reducers, Actions, and Action Creators.           |
| `appState.js`            | Sample application state maintained by Redux.                              |
| `store.js`               | Configuration and creation of the Redux store.                             |
| `types.ts`               | Typescript types used throughout the application.                          |

### Resources

1) [React/Redux Example App #1](https://github.com/erikras/react-redux-universal-hot-example/tree/master/src/redux)
2) [React/Redux Example App #2](https://github.com/goopscoop/ga-react-tutorial/tree/6-reduxActionsAndReducers/src/redux)
3) [React/Redux Example App #3](https://github.com/gothinkster/react-redux-realworld-example-app/tree/master/src)
4) [Redux Ducks Spec](https://github.com/erikras/ducks-modular-redux)
5) [React Router/Redux Setup Steps](https://github.com/supasate/connected-react-router#usage)
6) [Redux Hot Reloading](https://redux.js.org/recipes/configuring-your-store/#hot-reloading)
7) [Redux Thunk](https://github.com/reduxjs/redux-thunk)
8) [Redux/React Official Real World Example](https://github.com/reduxjs/redux/tree/master/examples/real-world)
