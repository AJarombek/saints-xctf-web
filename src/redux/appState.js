/**
 * Sample Redux State for the entire application.
 * @author Andrew Jarombek
 * @since 5/3/2020
 */

const appState = {
  auth: {
    auth: {
      isFetching: false,
      lastUpdated: 1588530309,
      signedIn: true,
      status: "SUCCESS"
    },
    registration: {
      stage: 1,
      first: "Andy",
      last: "Jarombek",
      email: "andrew@jarombek.com",
      username: undefined,
      password: undefined,
      activationCode: undefined,
      teams: {
        saintsXCTFAlumni: {
          status: "accepted",
          groups: {
            alumni: {
              status: "accepted"
            }
          }
        }
      }
    },
    user: {
      andy: {
        isFetching: false,
        didInvalidate: false,
        lastUpdated: 1588530310,
        username: "andy",
        first: "Andrew",
        last: "Jarombek",
        password: "xxxxxx",
        description: "Test Description",
        memberSince: "2016-02-23 12:00:00",
        classYear: 2017,
        location: "New York, NY",
        favoriteEvent: "Cooldown Jog",
        activationCode: "ABCD1234",
        email: "andrew@jarombek.com",
        lastSignIn: "2020-05-03 12:00:00",
        subscribed: "Y",
        logs: []
      }
    },
  },
  profile: {
    tom: {
      isFetching: false,
      didInvalidate: false,
      lastUpdated: 1588530311
    },
    joe: {
      isFetching: false,
      didInvalidate: false,
      lastUpdated: 1588530312
    },
    ben: {
      isFetching: false,
      didInvalidate: false,
      lastUpdated: 1588530313
    }
  },
  group: {
    wmenstf: {
      isFetching: false,
      didInvalidate: false,
      lastUpdated: 1588530314,
      groupName: "wmenstf",
      groupTitle: "Women's Track & Field",
      description: "",
      weekStart: "monday",
      members: [],
      logs: []
    }
  },
  logs: {
    isFetching: false,
    didInvalidate: false,
    lastUpdated: 1588530315,
    items: {
      1: {
        log_id: 1,
        username: "andy",
        first: "Andrew",
        last: "Jarombek",
        name: "Reservoir Loop",
        location: "New York, NY",
        date: "2020-05-03",
        type: "run",
        distance: 6.37,
        metric: "miles",
        miles: 6.37,
        time: "00:45:37",
        pace: "00:07:10",
        feel: 6,
        description: "",
        timeCreated: "2020-05-03 12:00:00",
        comments: {
          1: {
            commentId: 1,
            username: "andy",
            first: "Andrew",
            last: "Jarombek",
            logId: 1,
            time: "2020-05-03 12:00:00",
            content: "test"
          }
        }
      }
    }
  }
};
