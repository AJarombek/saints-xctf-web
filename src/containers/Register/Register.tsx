/**
 * Register component for people to create new accounts.
 * @author Andrew Jarombek
 * @since 5/9/2020
 */

import React, { useEffect, useRef, useState } from 'react';
import { useSelector } from 'react-redux';
import { useHistory, useLocation } from 'react-router-dom';
import queryString from 'query-string';

import { userAuthenticated } from '../../utils/auth';
import NavBar from '../../components/shared/NavBar';
import RegisterBody from '../../components/register/RegisterBody/RegisterBody';
import { RootState } from '../../redux/types';

type Props = {};

const Register: React.FunctionComponent<Props> = () => {
  const history = useHistory();
  const location = useLocation();

  const auth = useSelector((state: RootState) => state.auth.auth);
  const users = useSelector((state: RootState) => state.auth.user);
  const registration = useSelector((state: RootState) => state.registration);

  const [stage, setStage] = useState(registration.stage || 0);
  const ref = useRef(null);

  // In the development environment, stages can be skipped by adding the 'stage' query parameter
  // to the URL.
  if (process.env.NODE_ENV === 'local') {
    const parsed = queryString.parse(location.search);

    if (parsed.stage) {
      setStage(+parsed.stage);
    }
  }

  useEffect(() => {
    if (userAuthenticated(users, auth.signedInUser)) {
      history.push('/dashboard');
    }
  }, [users, auth.signedInUser, history]);

  useEffect(() => {
    setStage(registration.stage);
  }, [registration.stage]);

  return (
    <div className="sxctf-register" ref={ref}>
      <NavBar includeHeaders={['home', 'signIn', 'logo']} bodyRef={ref} />
      <RegisterBody stage={stage} />
    </div>
  );
};

export default Register;
