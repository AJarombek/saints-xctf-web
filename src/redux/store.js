/**
 * Build and configure the redux store, which holds the application state and handles
 * all state updates.
 * @author Andrew Jarombek
 * @since 5/1/2020
 */

import { createBrowserHistory } from 'history';
import { createStore, applyMiddleware } from 'redux';
import { routerMiddleware } from 'connected-react-router';
import createLogger from 'redux-logger';
import reducer from './modules/reducers';

const history = createBrowserHistory();

const configureStore = (data) => {
  const loggerMiddleware = createLogger();
  const middleware = [loggerMiddleware, routerMiddleware(history)];

  const createStoreWithMiddleware = applyMiddleware(...middleware)(createStore);
  return createStoreWithMiddleware(reducer, data);
};

export { history };
export default configureStore;
