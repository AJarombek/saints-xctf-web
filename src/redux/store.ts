/**
 * Build and configure the redux store, which holds the application state and handles
 * all state updates.
 * @author Andrew Jarombek
 * @since 5/1/2020
 */

import { createBrowserHistory } from 'history';
import { createStore, applyMiddleware, Store, Action } from 'redux';
import { routerMiddleware } from 'connected-react-router';
import { createLogger } from 'redux-logger';
import thunkMiddleware, { ThunkAction } from 'redux-thunk';
import reducer from './modules/reducers';
import { RootState } from './types';

export const history = createBrowserHistory();

export default function configureStore(): Store {
  const loggerMiddleware = createLogger();
  const middleware = [routerMiddleware(history), loggerMiddleware, thunkMiddleware];

  const createStoreWithMiddleware = applyMiddleware(...middleware)(createStore);
  const store = createStoreWithMiddleware(reducer(history));

  if (process.env.NODE_ENV !== 'production' && module.hot) {
    module.hot.accept('./modules/reducers', () => store.replaceReducer(reducer));
  }

  return store;
}

export type AppThunk<ReturnType = void, State = RootState> = ThunkAction<ReturnType, State, unknown, Action<string>>;
