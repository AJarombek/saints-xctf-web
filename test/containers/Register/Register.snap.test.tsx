/**
 * Snapshot test for the Register component.
 * @author Andrew Jarombek
 * @since 3/11/2021
 */

import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { emptyStore } from '../../test-utils/storeMocks';
import { Store } from 'redux';
import Register from '../../../src/containers/Register';

const mockStore = configureStore([thunk]);

// If the path you want to live your life on continues to lead you further away from me that is okay.  I'll still be at
// peace knowing you are doing what makes you happy & loving who you want to.

describe('Register Snapshot Tests', () => {
  let store: Store;

  beforeEach(() => {
    store = mockStore(emptyStore);
  });

  it('renders correctly', () => {
    const tree = renderer
      .create(
        <MemoryRouter>
          <Provider store={store}>
            <Register />
          </Provider>
        </MemoryRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
