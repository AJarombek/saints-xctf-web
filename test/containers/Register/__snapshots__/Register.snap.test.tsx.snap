// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Snapshot Tests renders correctly 1`] = `
<div
  className="sxctf-register"
>
  <div
    className="sxctf-nav-bar sxctf-nav-bar-dropdown-hidden dry-0-1-10 lightTheme-0-1-1 lightDropdownHeaderTheme-0-1-6"
  >
    <figure
      className="sxctf-logo"
    >
      <img
        alt=""
        onClick={[Function]}
        src="test-file-stub"
      />
    </figure>
    <h1
      onClick={[Function]}
    >
      SaintsXCTF
    </h1>
    <div
      className="sxctf-nav-buttons"
    >
      <div
        className="ajBaseButton-2-2-4 ajBaseButtonEnabled-2-2-5 aj-text-button ajTextButton-2-2-1 ajTextButtonEnabled-2-2-2 homeButton"
        onClick={[Function]}
      >
        <button
          disabled={false}
          type="button"
        >
          Home
        </button>
      </div>
      <div
        className="ajBaseButton-2-2-4 ajBaseButtonEnabled-2-2-5 aj-contained-button ajContainedButton-2-2-7 ajContainedButtonEnabled-2-2-8 signInButton"
        onClick={[Function]}
      >
        <button
          disabled={false}
          type="button"
        >
          Sign In
        </button>
      </div>
    </div>
    <div
      className="sxctf-nav-hamburger"
    >
      <div
        className="aj-mobile-hamburger ajMobileHamburger-2-2-10"
        onClick={[Function]}
      >
        <button
          type="button"
        >
          <span
            className="aj-mobile-hamburger-inactive ajMobileHamburgerInactive-2-2-12"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    className="sxctf-nav-dropdown sxctf-nav-dropdown-hidden dryDropdown-0-1-12 lightDropdownTheme-0-1-4"
  >
    <div
      className="ajNavList-2-2-13 aj-nav-list navList-0-1-13"
    >
      <div
        className="ajNavListItem-2-2-14 aj-nav-list-item"
      >
        <div
          onClick={[Function]}
        >
          <div>
            Home
          </div>
        </div>
      </div>
      <div
        className="ajNavListItem-2-2-14 aj-nav-list-item"
      >
        <div
          onClick={[Function]}
        >
          <div>
            Sign In
          </div>
        </div>
      </div>
      <div
        className="ajNavListItem-2-2-14 aj-nav-list-item ajNavListLastItem-2-2-15 aj-nav-list-last-item"
      >
        <div
          onClick={[Function]}
        >
          <div>
            <img
              alt=""
              className="mobile-dropdown-logo"
              src="test-file-stub"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="sxctf-register-body"
  >
    <div
      className="sxctf-register-personal-info"
    >
      <h2>
        Register
      </h2>
      <div
        className="form-inputs"
      >
        <div
          className="sxctf-image-input-set column"
        >
          <form>
            <div
              className="sxctf-image-input"
              data-cypress="image-input-firstName"
            >
              <div
                className="no-icon"
              >
                 
              </div>
              <input
                autoComplete="firstName"
                maxLength={30}
                name="firstName"
                onChange={[Function]}
                placeholder="First Name"
                type="text"
              />
              <div>
                <div
                  className="status none"
                >
                  <p />
                </div>
              </div>
            </div>
            <div
              className="sxctf-image-input"
              data-cypress="image-input-lastName"
            >
              <div
                className="no-icon"
              >
                 
              </div>
              <input
                autoComplete=""
                maxLength={30}
                name="lastName"
                onChange={[Function]}
                placeholder="Last Name"
                type="text"
              />
              <div>
                <div
                  className="status none"
                >
                  <p />
                </div>
              </div>
            </div>
          </form>
        </div>
        <div
          className="sxctf-image-input"
          data-cypress="image-input-email"
        >
          <img
            alt=""
            src="test-file-stub"
          />
          <input
            autoComplete=""
            maxLength={50}
            name="email"
            onChange={[Function]}
            placeholder="Email"
            type="text"
          />
          <div>
            <div
              className="status none"
            >
              <p />
            </div>
          </div>
        </div>
      </div>
      <div
        className="form-buttons"
      >
        <div
          className="ajBaseButton-2-2-4 ajBaseButtonDisabled-2-2-6 aj-contained-button ajContainedButton-2-2-7 ajContainedButtonDisabled-2-2-9"
          onClick={null}
        >
          <button
            disabled={true}
            type="button"
          >
            Continue
          </button>
        </div>
        <div
          className="ajBaseButton-2-2-4 ajBaseButtonEnabled-2-2-5 aj-text-button ajTextButton-2-2-1 ajTextButtonEnabled-2-2-2"
          onClick={[Function]}
        >
          <button
            disabled={false}
            type="button"
          >
            Exit
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
